sources: add Telegram source

atereshkin opened the pull request:
## Details

This implements Telegram social login provider as requested in #2232

Closes #2232

---

## Checklist

-   [x] Local tests pass (`ak test authentik/`)
-   [x] The code has been formatted (`make lint-fix`)

If an API change has been made

-   [x] The API schema has been updated (`make gen-build`)

If changes to the frontend have been made

-   [x] The code has been formatted (`make web`)

If applicable

-   [x] The documentation has been updated
-   [x] The documentation has been formatted (`make docs`)

atereshkin committed:
sources: add Telegram source (#2232)

atereshkin committed:
sources: add Telegram source (#2232)

atereshkin committed:
sources/telegram: put telegram user info into policy context (#2232)

atereshkin committed:
sources/telegram: replace regular input for bot token with a "secret" one (#2232)

atereshkin committed:
sources/telegram: fix typo on Telegram source form

atereshkin committed:
sources/telegram: added UserSourceConnection/GroupSourceConnection and SourceFlowManager subclasses for Telegram source

atereshkin committed:
sources/telegram: added UserSourceConnection/GroupSourceConnection and SourceFlowManager subclasses for Telegram source

atereshkin committed:
sources/telegram: improved code layout

atereshkin committed:
sources/telegram: collapsed migrations

atereshkin committed:
sources/telegram: collapsed migrations

atereshkin committed:
sources/telegram: fix lint errors

atereshkin committed:
sources/telegram: fix lint errors

atereshkin committed:
sources/telegram: fixed lint errors in docs

atereshkin committed:
sources/telegram: fix app config

tanberry commented on website/docs/developer-docs/index.md:
```diff
@@ -67,7 +67,8 @@ authentik
 │   ├── oauth - OAuth1 and OAuth2 Source
 │   ├── plex - Plex source
-│   └── saml - SAML2 Source
+│   ├── saml - SAML2 Source
```
Both "source" and "provider" should be lower-case... I see we have long had a mix. @atereshkin would you please fix the others, too? :-)

tanberry commented on website/docs/sidebar.mjs:
```diff
@@ -568,8 +568,9 @@ const items = [
-                            "users-sources/sources/social-logins/twitch/index",
                             "users-sources/sources/social-logins/plex/index",
+                            "users-sources/sources/social-logins/telegram/index",
+                            "users-sources/sources/social-logins/twitch/index",
```
Thanks for fixing the wrong alphabetization for Twitch!

tanberry commented on website/docs/users-sources/sources/social-logins/telegram/index.md:
```diff
@@ -0,0 +1,52 @@
+support_level: community
+---
+
+Allows users to authenticate using their Telegram account via Telegram application
```
```suggestion
Configuring Telegram as a source allows users to authenticate using their Telegram account credentials within authentik.
```

tanberry commented on website/docs/users-sources/sources/social-logins/telegram/index.md:
```diff
@@ -0,0 +1,52 @@
+
+## Telegram
+
+To use Telegram as a source, you need to register a Telegram bot first:
```
```suggestion
To use Telegram as a source, you first need to register a Telegram bot:
```

tanberry commented on website/docs/users-sources/sources/social-logins/telegram/index.md:
```diff
@@ -0,0 +1,52 @@
+
+## authentik
+
+In authentik, go to _Directory_ > _Federation & Social login_ and click _Create_.
```
For UI elements we use bold. Italics are only used for emphasis.

tanberry commented on website/docs/users-sources/sources/social-logins/telegram/index.md:
```diff
@@ -0,0 +1,52 @@
+In authentik, go to _Directory_ > _Federation & Social login_ and click _Create_.
+Select _Telegram source_ and configure your source:
+
+- _Name:_ Choose a name
```
```suggestion
- **Name**: Choose a name
```

rissson commented on authentik/sources/telegram/models.py:
```diff
@@ -0,0 +1,138 @@
+class TelegramSource(Source):
+    """Log in with Telegram."""
+
+    bot_username = models.CharField(max_length=255, help_text=_("Telegram bot username"))
+    bot_token = models.CharField(max_length=255, help_text=_("Telegram bot token"))
```
Yeah we usually don't bother and just use TextField, it's the same underlying data type in postgres anyway, so performance is not a concern.

atereshkin replied:
done

tanberry commented on website/docs/users-sources/sources/social-logins/telegram/index.md:
```diff
@@ -0,0 +1,52 @@
+Select _Telegram source_ and configure your source:
+
+- _Name:_ Choose a name
+- _Slug:_ Set a slug
```
```suggestion
- **Slug**: Set a slug
```

tanberry commented on website/docs/users-sources/sources/social-logins/telegram/index.md:
```diff
@@ -0,0 +1,52 @@
+
+- `info`: A Python dictionary containing Telegram user data:
+    - `id` - Telegram user ID
+    - `username` - Username of the user. May not be present
```
```suggestion
    - `username` - Username of the user. Might not be present.
```

tanberry commented on website/docs/users-sources/sources/social-logins/telegram/index.md:
```diff
@@ -0,0 +1,52 @@
+- `info`: A Python dictionary containing Telegram user data:
+    - `id` - Telegram user ID
+    - `username` - Username of the user. May not be present
+    - `first_name` - First name of the user. May not be present
```
```suggestion
    - `first_name` - First name of the user. Might not be present.
```

tanberry commented on website/docs/users-sources/sources/social-logins/telegram/index.md:
```diff
@@ -0,0 +1,52 @@
+    - `id` - Telegram user ID
+    - `username` - Username of the user. May not be present
+    - `first_name` - First name of the user. May not be present
+    - `last_name` - Last name of the user. May not be present
```
```suggestion
    - `last_name` - Last name of the user. Might not be present.
```

tanberry reviewed (changes requested):
Thanks so much @atereshkin for adding the docs for Telegram. A few nits to address please, but looks great, LGTM!!

tanberry commented:
Oh, also @atereshkin in order to resolve some of the build checks fails, you might need to 1. rebase your local repo on upstream `main` in order to pick up some changes in our build tools, and 2. run `make docs-install` to instlal those new tools/versions.

atereshkin committed:
sources/telegram: add user source settings UI so that the users can disconnect Telegram source from their account

atereshkin committed:
sources/telegram: clean up code per @risson's suggestions

atereshkin committed:
sources/telegram: clean up code per @risson's suggestions

atereshkin committed:
Update website/docs/users-sources/sources/social-logins/telegram/index.md

Co-authored-by: Tana M Berry <tanamarieberry@yahoo.com>
Signed-off-by: Alexander Tereshkin <96586+atereshkin@users.noreply.github.com>

atereshkin committed:
Update website/docs/users-sources/sources/social-logins/telegram/index.md

Co-authored-by: Tana M Berry <tanamarieberry@yahoo.com>
Signed-off-by: Alexander Tereshkin <96586+atereshkin@users.noreply.github.com>

atereshkin committed:
Update website/docs/users-sources/sources/social-logins/telegram/index.md

Co-authored-by: Tana M Berry <tanamarieberry@yahoo.com>
Signed-off-by: Alexander Tereshkin <96586+atereshkin@users.noreply.github.com>

atereshkin committed:
Update website/docs/users-sources/sources/social-logins/telegram/index.md

Co-authored-by: Tana M Berry <tanamarieberry@yahoo.com>
Signed-off-by: Alexander Tereshkin <96586+atereshkin@users.noreply.github.com>

atereshkin committed:
Update website/docs/users-sources/sources/social-logins/telegram/index.md

Co-authored-by: Tana M Berry <tanamarieberry@yahoo.com>
Signed-off-by: Alexander Tereshkin <96586+atereshkin@users.noreply.github.com>

atereshkin committed:
Update website/docs/users-sources/sources/social-logins/telegram/index.md

Co-authored-by: Tana M Berry <tanamarieberry@yahoo.com>
Signed-off-by: Alexander Tereshkin <96586+atereshkin@users.noreply.github.com>

atereshkin committed:
Update website/docs/users-sources/sources/social-logins/telegram/index.md

Co-authored-by: Tana M Berry <tanamarieberry@yahoo.com>
Signed-off-by: Alexander Tereshkin <96586+atereshkin@users.noreply.github.com>

atereshkin committed:
Update website/docs/users-sources/sources/social-logins/telegram/index.md

Co-authored-by: Tana M Berry <tanamarieberry@yahoo.com>
Signed-off-by: Alexander Tereshkin <96586+atereshkin@users.noreply.github.com>

atereshkin committed:
Update website/docs/users-sources/sources/social-logins/telegram/index.md

Co-authored-by: Tana M Berry <tanamarieberry@yahoo.com>
Signed-off-by: Alexander Tereshkin <96586+atereshkin@users.noreply.github.com>

atereshkin committed:
Update website/docs/users-sources/sources/social-logins/telegram/index.md

Co-authored-by: Tana M Berry <tanamarieberry@yahoo.com>
Signed-off-by: Alexander Tereshkin <96586+atereshkin@users.noreply.github.com>

atereshkin committed:
Update website/docs/users-sources/sources/social-logins/telegram/index.md

Co-authored-by: Tana M Berry <tanamarieberry@yahoo.com>
Signed-off-by: Alexander Tereshkin <96586+atereshkin@users.noreply.github.com>

atereshkin committed:
Update website/docs/users-sources/sources/social-logins/telegram/index.md

Co-authored-by: Tana M Berry <tanamarieberry@yahoo.com>
Signed-off-by: Alexander Tereshkin <96586+atereshkin@users.noreply.github.com>

atereshkin committed:
Update website/docs/users-sources/sources/social-logins/telegram/index.md

Co-authored-by: Tana M Berry <tanamarieberry@yahoo.com>
Signed-off-by: Alexander Tereshkin <96586+atereshkin@users.noreply.github.com>

atereshkin committed:
Update website/docs/users-sources/sources/social-logins/telegram/index.md

Co-authored-by: Tana M Berry <tanamarieberry@yahoo.com>
Signed-off-by: Alexander Tereshkin <96586+atereshkin@users.noreply.github.com>

atereshkin committed:
sources/telegram: improve docs based on @tanberry's suggestions

atereshkin commented on authentik/sources/telegram/models.py:
```diff
@@ -0,0 +1,138 @@
+from authentik.flows.challenge import RedirectChallenge
+
+
+class TelegramSourcePropertyMapping(PropertyMapping):
```
done

atereshkin commented on authentik/sources/telegram/models.py:
```diff
@@ -0,0 +1,138 @@
+
+    bot_username = models.CharField(max_length=255, help_text=_("Telegram bot username"))
+    bot_token = models.CharField(max_length=255, help_text=_("Telegram bot token"))
+    request_access = models.BooleanField(
```
done

atereshkin committed:
sources/telegram: fix minor docs formatting issue

atereshkin committed:
sources/teleram: add tests for views

tanberry reviewed (approved):
Thanks @atereshkin !

atereshkin commented on authentik/sources/telegram/stage.py:
```diff
@@ -0,0 +1,48 @@
+class TelegramChallengeResponse(ChallengeResponse):
+    component = CharField(default="ak-source-telegram")
+
+    id = serializers.IntegerField()
```
That's Telegram user id. I just named these fields the same way Telegram calls them. Do you think we should change it to something that makes sense in the context of Authentik or a simple comment would suffice?

atereshkin committed:
sources/telegram: update serielizer field types references to be in line with convention

atereshkin committed:
sources/telegram: add missing type annotations

atereshkin committed:
sources/telegram: add check for source.enabled in the redirect view

BeryJu commented on authentik/sources/telegram/views.py:
```diff
@@ -0,0 +1,68 @@
+class TelegramStartView(View):
+    def get(self, request, source_slug):
+        source = TelegramSource.objects.get(slug=source_slug)
+        plan: FlowPlan = self.request.session[SESSION_KEY_PLAN]
```
One solution for this that would work is similarly to the SAML Source: https://github.com/goauthentik/authentik/blob/main/authentik/sources/saml/views.py#L67-L140

Add a pre-authentication flow to the model, and in this view create a plan for that flow and append the telegram stage to it

BeryJu replied:
The idea would be that the sources' `ui_login_button` returns a `RedirectChallenge` to this view that initiates the pre_authorization flow and does `plan.append_stage(in_memory_stage(TelegramLoginView))` (which would allow user configured stages to run before telegram. Then after the authentication is done we use the default backend flow logic with either continuing with the authentication or enrollment flow

atereshkin committed:
sources/telegram: add pre-authentication flow to telegram source

atereshkin committed:
sources/teregram: fix merge errors

atereshkin committed:
sources/telegram: improve docs wording

dewi-tik committed:
Standardized documentation

dewi-tik commented:
Hope you don't mind, I've edited the documentation to better match our current style guide. Great PR. Looking forward to getting it merged.

dependabot[bot] committed:
web: bump @types/node from 22.15.19 to 24.5.2 in /web (#16989)

Bumps [@types/node](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/HEAD/types/node) from 22.15.19 to 24.5.2.
- [Release notes](https://github.com/DefinitelyTyped/DefinitelyTyped/releases)
- [Commits](https://github.com/DefinitelyTyped/DefinitelyTyped/commits/HEAD/types/node)

---
updated-dependencies:
- dependency-name: "@types/node"
  dependency-version: 24.5.2
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

atereshkin committed:
sources/telegram: added telegram source package to the list of ignored modules for mypy

atereshkin committed:
sources/telegram: fix TS lint errors

atereshkin committed:
sources/telegram: improve test coverage
